File: packages/network_info_plus/network_info_plus/android/src/main/java/dev/fluttercommunity/plus/network_info/NetworkInfo.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package dev.fluttercommunity.plus.network_info;

import android.annotation.SuppressLint;
import android.net.DhcpInfo;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;
import java.util.List;

/** Reports network info such as wifi name and address. */
class NetworkInfo {
  private final WifiManager wifiManager;

  NetworkInfo(WifiManager wifiManager) {
    this.wifiManager = wifiManager;
  }

  String getWifiName() {
    WifiInfo wifiInfo = getWifiInfo();
    String ssid = null;
    if (wifiInfo != null) ssid = wifiInfo.getSSID();
    if (ssid != null) ssid = ssid.replaceAll("\"", ""); // Android returns "SSID"
    return ssid;
  }

  String getWifiBSSID() {
    WifiInfo wifiInfo = getWifiInfo();
    String bssid = null;
    if (wifiInfo != null) {
      bssid = wifiInfo.getBSSID();
    }
    return bssid;
  }

  @SuppressLint("DefaultLocale")
  String getWifiIPAddress() {
    WifiInfo wifiInfo = null;
    if (wifiManager != null) wifiInfo = wifiManager.getConnectionInfo();
    String ip = null;
    int i_ip = 0;
    if (wifiInfo != null) i_ip = wifiInfo.getIpAddress();

    if (i_ip != 0)
      ip =
          String.format(
              "%d.%d.%d.%d",
              (i_ip & 0xff), (i_ip >> 8 & 0xff), (i_ip >> 16 & 0xff), (i_ip >> 24 & 0xff));

    return ip;
  }

  String getWifiSubnetMask() {
    String ip = this.getWifiIPAddress();
    String subnet = "";
    try {
      InetAddress inetAddress = InetAddress.getByName(ip);
      subnet = getIPv4Subnet(inetAddress);
    } catch (Exception ignored) {
    }
    return subnet;
  }

  String getBroadcast() {
    String broadcastIP = null;
    String ip = this.getWifiIPAddress();
    try {
      NetworkInterface ni = NetworkInterface.getByInetAddress(InetAddress.getByName(ip));
      for (InterfaceAddress address : ni.getInterfaceAddresses()) {
        if (!address.getAddress().isLoopbackAddress()) {
          InetAddress broadCast = address.getBroadcast();
          if (broadCast != null) {
            broadcastIP = broadCast.toString();
          }
        }
      }
    } catch (Exception ignored) {
    }
    return broadcastIP;
  }

  public String getIpV6() {
    try {
      String ip = this.getWifiIPAddress();
      NetworkInterface ni = NetworkInterface.getByInetAddress(InetAddress.getByName(ip));
      for (InterfaceAddress interfaceAddress : ni.getInterfaceAddresses()) {
        InetAddress address = interfaceAddress.getAddress();
        if (!address.isLoopbackAddress() && address instanceof Inet6Address) {
          String ipaddress = address.getHostAddress();
          if (ipaddress != null) {
            return ipaddress.split("%")[0];
          }
        }
      }
    } catch (Exception ignored) {
    }
    return null;
  }

  String getGatewayIpAdress() {
    DhcpInfo dhcpInfo = this.wifiManager.getDhcpInfo();
    int gatewayIPInt = dhcpInfo.gateway;
    @SuppressLint("DefaultLocale")
    String gatewayIP =
        String.format(
            "%d.%d.%d.%d",
            ((gatewayIPInt) & 0xFF),
            ((gatewayIPInt >> 8) & 0xFF),
            ((gatewayIPInt >> 16) & 0xFF),
            ((gatewayIPInt >> 24) & 0xFF));
    return gatewayIP;
  }

  private WifiInfo getWifiInfo() {
    return wifiManager == null ? null : wifiManager.getConnectionInfo();
  }

  private String getIPv4Subnet(InetAddress inetAddress) {
    try {
      NetworkInterface ni = NetworkInterface.getByInetAddress(inetAddress);
      List<InterfaceAddress> intAddrs = ni.getInterfaceAddresses();
      for (InterfaceAddress ia : intAddrs) {
        if (!ia.getAddress().isLoopbackAddress() && ia.getAddress() instanceof Inet4Address) {
          final InetAddress networkPrefix =
              getIPv4SubnetFromNetPrefixLength(ia.getNetworkPrefixLength());
          if (networkPrefix != null) {
            return networkPrefix.getHostAddress();
          }
        }
      }
    } catch (Exception ignored) {
    }
    return "";
  }

  private InetAddress getIPv4SubnetFromNetPrefixLength(int netPrefixLength) {
    try {
      int shift = (1 << 31);
      for (int i = netPrefixLength - 1; i > 0; i--) {
        shift = (shift >> 1);
      }
      String subnet =
          ((shift >> 24) & 255)
              + "."
              + ((shift >> 16) & 255)
              + "."
              + ((shift >> 8) & 255)
              + "."
              + (shift & 255);
      return InetAddress.getByName(subnet);
    } catch (Exception ignored) {
    }
    return null;
  }
}

File: packages/network_info_plus/network_info_plus/android/src/main/kotlin/dev/fluttercommunity/plus/network_info/NetworkInfo.kt
package dev.fluttercommunity.plus.network_info

import android.net.ConnectivityManager
import android.net.wifi.WifiInfo
import android.net.wifi.WifiManager
import android.os.Build
import java.net.*

/** Reports network info such as wifi name and address. */
internal class NetworkInfo(
    private val wifiManager: WifiManager,
    private val connectivityManager: ConnectivityManager? = null
) {

    // Using deprecated `connectionInfo` call here to be able to get info on demand
    @Suppress("DEPRECATION")
    private val wifiInfo: WifiInfo
        get() = wifiManager.connectionInfo

    // Android returns "SSID"
    fun getWifiName(): String? = wifiInfo.ssid

    fun getWifiBSSID(): String? = wifiInfo.bssid

    fun getWifiIPAddress(): String? {
        var ipAddress: String? = null

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val linkAddresses =
                connectivityManager?.getLinkProperties(connectivityManager.activeNetwork)?.linkAddresses

            val ipV4Address = linkAddresses?.firstOrNull { linkAddress ->
                linkAddress.address.hostAddress?.contains('.')
                    ?: false
            }?.address?.hostAddress

            ipAddress = ipV4Address
        } else {
            @Suppress("DEPRECATION")
            val interfaceIp = wifiInfo.ipAddress
            if (interfaceIp != 0) ipAddress = formatIPAddress(interfaceIp)
        }
        return ipAddress
    }

    fun getWifiSubnetMask(): String {
        val ip = getWifiIPAddress()
        val inetAddress = InetAddress.getByName(ip)
        val subnet = getIPv4Subnet(inetAddress)
        return subnet
    }

    fun getBroadcastIP(): String? {
        val currentWifiIpAddress = getWifiIPAddress()
        val inetAddress = InetAddress.getByName(currentWifiIpAddress)
        val networkInterface = NetworkInterface.getByInetAddress(inetAddress)
        networkInterface.interfaceAddresses.forEach { interfaceAddress ->
            if (!interfaceAddress.address.isLoopbackAddress) {
                if (interfaceAddress.broadcast != null) {
                    return interfaceAddress.broadcast.hostAddress
                }
            }
        }
        return null
    }

    fun getIpV6(): String? {
        val ip = getWifiIPAddress()
        val ni = NetworkInterface.getByInetAddress(InetAddress.getByName(ip))
        for (interfaceAddress in ni.interfaceAddresses) {
            val address = interfaceAddress.address
            if (!address.isLoopbackAddress && address is Inet6Address) {
                val ipaddress = address.getHostAddress()
                if (ipaddress != null) {
                    return ipaddress.split("%").toTypedArray()[0]
                }
            }
        }
        return null
    }

    fun getGatewayIPAddress(): String? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val linkAddresses =
                connectivityManager?.getLinkProperties(connectivityManager.activeNetwork)
            val dhcpServer = linkAddresses?.dhcpServerAddress?.hostAddress

            dhcpServer
        } else {
            @Suppress("DEPRECATION")
            val dhcpInfo = wifiManager.dhcpInfo
            val gatewayIPInt = dhcpInfo?.gateway

            gatewayIPInt?.let { formatIPAddress(it) }
        }
    }

    private fun formatIPAddress(intIP: Int): String =
        String.format(
            "%d.%d.%d.%d",
            intIP and 0xFF,
            intIP shr 8 and 0xFF,
            intIP shr 16 and 0xFF,
            intIP shr 24 and 0xFF
        )

    private fun getIPv4Subnet(inetAddress: InetAddress): String {
        val ni = NetworkInterface.getByInetAddress(inetAddress)
        val intAddresses = ni.interfaceAddresses
        for (ia in intAddresses) {
            if (!ia.address.isLoopbackAddress && ia.address is Inet4Address) {
                val networkPrefix =
                    getIPv4SubnetFromNetPrefixLength(ia.networkPrefixLength.toInt())
                if (networkPrefix != null) {
                    return networkPrefix.hostAddress!!
                }
            }
        }
        return ""
    }

    private fun getIPv4SubnetFromNetPrefixLength(netPrefixLength: Int): InetAddress? {
        var shift = 1 shl 31
        for (i in netPrefixLength - 1 downTo 1) {
            shift = shift shr 1
        }
        val subnet = ((shift shr 24 and 255)
            .toString() + "."
            + (shift shr 16 and 255)
            + "."
            + (shift shr 8 and 255)
            + "."
            + (shift and 255))
        return InetAddress.getByName(subnet)
    }
}
